src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,186 @@
+coefficients come from the semiring `R` and the variable `T` commutes with everything.
+
+Since we are going to convert back and forth between polynomials and Laurent polynomials, we
+decided to maintain some distinction by using the symbol `T`, rather than `X`, as the variable for
+Laurent polynomials
+
+## Notation
+The symbol `R[T;T⁻¹]` stands for `laurent_polynomial R`.  We also define
+
+* `C : R →+* R[T;T⁻¹]` the inclusion of constant polynomials, analogous to the one for `R[X]`;
+* `T : ℤ → R[T;T⁻¹]` the sequence of powers of the variable `T`.
+
+## Implementation notes
+
+We define Laurent polynomials as `add_monoid_algebra R ℤ`.
+Thus, they are essentially `finsupp`s `ℤ →₀ R`.
+This choice differs from the current irreducible design of `polynomial`, that instead shields away
+the implementation via `finsupp`s.  It is closer to the original definition of polynomials.
+
+As a consequence, `laurent_polynomial` plays well with polynomials, but there is a little roughness
+in establishing the API, since the `finsupp` implementation of `polynomial R` is well-shielded.
+
+Unlike the case of polynomials, I felt that the exponent notation was not too easy to use, as only
+natural exponents would be allowed.  Moreover, in the end, it seems likely that we should aim to
+perform computations on exponents in `ℤ` anyway and separating this via the symbol `T` seems
+convenient.
+
+I made a *heavy* use of `simp` lemmas, aiming to bring Laurent polynomials to the form `C a * T n`.
+Any comments or suggestions for improvements is greatly appreciated!
+
+##  Future work
+Lots is missing!  I would certainly like to show that `R[T;T⁻¹]` is the localization of `R[X]`
+inverting `X`.  This should be mostly in place, given `exists_T_pow` (which is part of PR #13415).
+-/
+
+open_locale polynomial big_operators
+open polynomial add_monoid_algebra finsupp
+noncomputable theory
+
+variables {R : Type*}
+
+/--  The semiring of Laurent polynomials with coefficients in the semiring `R`.
+We denote it by `R[T;T⁻¹]`.
+The ring homomorphism `C : R →+* R[T;T⁻¹]` includes `R` as the constant polynomials. -/
+abbreviation laurent_polynomial (R : Type*) [semiring R] := add_monoid_algebra R ℤ
+
+local notation R`[T;T⁻¹]`:9000 := laurent_polynomial R
+
+/--  The ring homomorphism, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent [semiring R] : R[X] →+* R[T;T⁻¹] :=
+(map_domain_ring_hom R int.of_nat_hom).comp (to_finsupp_iso R)
+
+/-- This is not a simp lemma, as it is usually preferable to use the lemmas about `C` and `X`
+instead. -/
+lemma polynomial.to_laurent_apply [semiring R] (p : R[X]) :
+  p.to_laurent = p.to_finsupp.map_domain coe := rfl
+
+/--  The `R`-algebra map, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent_alg [comm_semiring R] :
+  R[X] →ₐ[R] R[T;T⁻¹] :=
```

Comment:
Under what conditions is it an alg_hom over polynomial R itself?

Reply:
The algebra structure that statement would need is dangerous wrt instance diamonds

Reply:
Yakov, if you look at lines 336+ of the bigger Laurent PR, you will find the instance, commented out, for the algebra structure that you mention.  `R` is a `comm_semiring`, since some commutativity is required by `algebra`.

As Eric says, though, this may cause problems for mathlib.

Reply:
I ran into something similar when trying to work on the poly -> power_series -> laurent_series -> ratfunc hierachy. One way to generalize here would be the make an additional `→ₐ[S]` alg_hom where `[algebra S R] [algebra S R[X]] [algebra S R[T;T⁻¹]`, if I remember correctly.

Reply:
Something similar happens already at the level of `add_monoid_algebra`s, where there are `non_commutative_associative_unital` versions of some of the maps.  I think that if that is the kind of direction that someone wants, then probably working directly with `add_monoid_algebra`s might be better.  If you have a different opinion, feel free to say so and I can add this!